### ISSUE TITLE
HTTP2 backend connections not beeing reused with multiple HTTP2 streams in single connexion

### ISSUE THREAD

**Issue by crohart** [haproxy/haproxy#1442]
### Detailed Description of the Problem

Backend connection persistence duning the 'pool-purge-delay' does not seem to work when using HTTP2 backends and http-reuse policy Never. 
We see a new HTTP2 backend connection established for every new requests where it should not.
It seams like the backend connection is re-opened for every new H2 stream in the same frontend connection. 
In v2.0, it was working fine and new streams in the same upfront connection where correctly re-using the backend connection with reuse Never.

It worked and still works with Haproxy 2.0.x on my HA servers
It does not work with any sub-version in the Haproxy 2.2.x branch.

### Expected Behavior

The Haproxy should persist the backend HTTP2 connection as long the backend servers implement keepalive correctly.

### Steps to Reproduce the Behavior

configure 2 backends with HTTP2 and ALPN HTTPS only.
configure the HTTP-Reuse policy to Never.

Open a HTTP2 connection to the frontend 
Send a first stream request -> it creates one backend stream over a TCP/SSL connection to one of the backends.
Close the stream but keep the upfront connection up.
Send a second stream request -> it does not reuse the TCP/SSL connection established in the first req and re-establishes a new connection.

Seems like only one stream is allowed in one backend connections...
In v2.0 we could have one connection with n stream.
But in v2.2 we can only achieve one stream per connection because it closes the connection immediately when we close one stream.

### Do you have any idea what may have caused this?

probably due to an optimization in versions 2.2.x (and may be v2.1 - not tested)

### What is your configuration?

```haproxy
Haproxy 2.2.17 running on OPNSense
```

### Output of `haproxy -vv`

```plain
HA-Proxy version 2.2.17-dd94a25 2021/09/07 - https://haproxy.org/
Status: long-term supported branch - will stop receiving fixes around Q2 2025.
Known bugs: http://www.haproxy.org/bugs/bugs-2.2.17.html
Running on: FreeBSD 12.1-RELEASE-p20-HBSD FreeBSD 12.1-RELEASE-p20-HBSD #0  4dbd2ffa241(stable/21.7)-dirty: Tue Oct 26 08:46:43 CEST 2021     root@sensey:/usr/obj/usr/src/amd64.amd64/sys/SMP amd64
Build options :
  TARGET  = freebsd
  CPU     = generic
  CC      = cc
  CFLAGS  = -O2 -pipe -DHARDENEDBSD -fstack-protector-strong -fno-strict-aliasing -Wall -Wextra -Wdeclaration-after-statement -fwrapv -Wno-address-of-packed-member -Wno-unused-label -Wno-sign-compare -Wno-unused-parameter -Wno-missing-field-initializers -Wno-string-plus-int -Wtype-limits -Wshift-negative-value -Wnull-dereference -DFREEBSD_PORTS
  OPTIONS = USE_PCRE=1 USE_PCRE_JIT=1 USE_STATIC_PCRE=1 USE_GETADDRINFO=1 USE_OPENSSL=1 USE_LUA=1 USE_ACCEPT4=1 USE_ZLIB=1 USE_CPU_AFFINITY=1
  DEBUG   = 

Feature list : -EPOLL +KQUEUE -NETFILTER +PCRE +PCRE_JIT -PCRE2 -PCRE2_JIT +POLL -PRIVATE_CACHE +THREAD -PTHREAD_PSHARED -BACKTRACE +STATIC_PCRE -STATIC_PCRE2 +TPROXY -LINUX_TPROXY -LINUX_SPLICE +LIBCRYPT -CRYPT_H +GETADDRINFO +OPENSSL +LUA -FUTEX +ACCEPT4 +CLOSEFROM +ZLIB -SLZ +CPU_AFFINITY -TFO -NS -DL -RT -DEVICEATLAS -51DEGREES -WURFL -SYSTEMD -OBSOLETE_LINKER -PRCTL -THREAD_DUMP -EVPORTS

Default settings :
  bufsize = 16384, maxrewrite = 1024, maxpollevents = 200

Built with multi-threading support (MAX_THREADS=64, default=2).
Built with OpenSSL version : OpenSSL 1.1.1l  24 Aug 2021
Running on OpenSSL version : OpenSSL 1.1.1l  24 Aug 2021
OpenSSL library supports TLS extensions : yes
OpenSSL library supports SNI : yes
OpenSSL library supports : TLSv1.0 TLSv1.1 TLSv1.2 TLSv1.3
Built with Lua version : Lua 5.3.6
Built with zlib version : 1.2.11
Running on zlib version : 1.2.11
Compression algorithms supported : identity("identity"), deflate("deflate"), raw-deflate("deflate"), gzip("gzip")
Built with transparent proxy support using: IP_BINDANY IPV6_BINDANY
Built with PCRE version : 8.45 2021-06-15
Running on PCRE version : 8.45 2021-06-15
PCRE library supports JIT : no (libpcre build without JIT?)
Encrypted password support via crypt(3): yes
Built with clang compiler version 8.0.1 (tags/RELEASE_801/final 366581)
Built with the Prometheus exporter as a service

Available polling systems :
     kqueue : pref=300,  test result OK
       poll : pref=200,  test result OK
     select : pref=150,  test result OK
Total: 3 (3 usable), will use kqueue.

Available multiplexer protocols :
(protocols marked as <default> cannot be specified using 'proto' keyword)
            fcgi : mode=HTTP       side=BE        mux=FCGI
       <default> : mode=HTTP       side=FE|BE     mux=H1
              h2 : mode=HTTP       side=FE|BE     mux=H2
       <default> : mode=TCP        side=FE|BE     mux=PASS

Available services : prometheus-exporter
Available filters :
	[SPOE] spoe
	[COMP] compression
	[TRACE] trace
	[CACHE] cache
	[FCGI] fcgi-app
```

**Comment by wtarreau** [comment]
Thanks for your report. At first glance I don't see any reason for this to happen but it's possible that we broke something. When "http-reuse" is set to "never", normally the idle connections are attached to the front session so that subsequent requests can find it there. We'll have a look.

**Comment by wtarreau** [comment]
By the way, could you provide a simple reproducer config ? Usually when there's no conf, we create one ourselves just to figure that with our conf it works...  Also it could be great if you could give a try to 2.4 (2.4.8 at the moment of writing), which has all known bugs fixed.

**Comment by crohart** [comment]
It's the following simple config where our Haproxies are forwarding an HTTP2-only API to our backend workers.
It's not used for Web but HTTP2 RPC requests.
The backend Workers are based on Apache and use HTTP2-only. The Workers have keepalives working fine. 
And if we do not proxy the requests it also work fine too.

Again it works as long as we replace the default opnsense haproxy binary (ie. the 2.2) with the 2.0 versions.
With the 2.2, our Workers receive new connections from Haproxies for every new API requests from the same client in the same established upfront connection.

I can't test the 2.4 easily as this configs are running on OPNSense appliances (latest versions).

Btw- Is there anything that changed between 2.0 and 2.2 regarding the backend connection pooling - that could have such impact? 

Below the config:

```
#
# Automatically generated configuration.
# Do not edit this file manually.
#

global
    uid                         80
    gid                         80
    chroot                      /var/haproxy
    daemon
    stats                       socket /var/run/haproxy.socket group proxy mode 775 level admin
    nbproc                      1
    nbthread                    8
    maxconn                     25000
    tune.ssl.default-dh-param   1024
    spread-checks               0
    tune.chksize                16384
    tune.bufsize                16384
    tune.lua.maxmem             0
    log /var/run/log local0 warning
    ssl-default-bind-options no-tls-tickets ssl-min-ver TLSv1.2 ssl-max-ver TLSv1.3
    ssl-default-bind-ciphers ECDHE-ECDSA-AES256-GCM-SHA384:ECDHE-RSA-AES256-GCM-SHA384:ECDHE-ECDSA-CHACHA20-POLY1305:ECDHE-RSA-CHACHA20-POLY1305:ECDHE-ECDSA-AES128-GCM-SHA256:ECDHE-RSA-AES128-GCM-SHA256:ECDHE-ECDSA-AES256-SHA384:ECDHE-RSA-AES256-SHA384:ECDHE-ECDSA-AES128-SHA256:ECDHE-RSA-AES128-SHA256
    ssl-default-bind-ciphersuites TLS_AES_256_GCM_SHA384:TLS_CHACHA20_POLY1305_SHA256:TLS_AES_128_GCM_SHA256

defaults
    log     global
    option redispatch -1
    timeout client 30s
    timeout connect 30s
    timeout server 30s
    retries 3
    default-server init-addr last,libc

# autogenerated entries for ACLs


# autogenerated entries for config in backends/frontends

# autogenerated entries for stats

# Frontend: Main_Services_Prod ()
frontend Main_Services_Prod
    bind 0.0.0.0:443 name 0.0.0.0:443 ssl alpn h2,http/1.1 crt-list /tmp/haproxy/ssl/5fc20302e5b717.03896565.certlist 
    bind :::443 name :::443 ssl alpn h2,http/1.1 crt-list /tmp/haproxy/ssl/5fc20302e5b717.03896565.certlist 
    mode http
    option http-keep-alive
    option forwardfor
    # tuning options
    timeout client 30s

    # logging options
    # ACL: APIv1_Path
    acl acl_5fc208c284cac9.57267115 path_beg -i /api-v1

    # ACTION: APIv1_Backend_Prod
    use_backend APIv1_Prod if acl_5fc208c284cac9.57267115


# Backend: APIv1_Prod ()
backend APIv1_Prod
    option log-health-checks
    # health check: TCP_Check
    mode http
    balance source

    # tuning options
    timeout connect 30s
    timeout server 30s
    http-reuse never
    server Worker1 10.0.0.8:443 check inter 2s ssl alpn h2 verify required ca-file /tmp/haproxy/ssl/5fc203585465c7.88358403.calist
    server Worker2 10.0.0.9:443 check inter 2s ssl alpn h2 verify required ca-file /tmp/haproxy/ssl/5fc20368667019.91138898.calist


listen local_statistics
    bind            127.0.0.1:8822
    mode            http
    stats uri       /haproxy?stats
    stats realm     HAProxy\ statistics
    stats admin     if TRUE

# statistics are DISABLED

```

**Comment by wtarreau** [comment]
Thanks, we'll test with that. Yes there were changes, the most important ones being that in 2.0 the connections are tied to a thread, and when you have many threads, your number of idle connections explode, or you can't keep them open. In 2.2 they're shared between all threads so this problem doesn't happen anymore.

If you're curious and do not have too much traffic, you can enable traces on the CLI at the H2 level. Be careful, this can be quite verbose, so you'll need to send that to a file:
```
$ (echo "trace h2 sink buf0; trace h2 level developer; trace h2 verbosity complete; trace h2 start now;show events buf0 -w -n" ; cat ) | socat - /path/to/stats-socket
```

In order to stop the traces (as even if you don't read them they continue to be produced in a ring buffer):
```
$ socat - /path/to/stats-socket <<< "trace 0"
```

The trace will contain all watch points and decisions taken in the code for a connection. You may discover there some GOAWAY frames indicating that there might be another factor for these connections to be closed. But I suspect it's entirely related to the idle conns instead.

**Comment by zawachte-c** [comment]
Any updates on this issue? Do we know if this repros on 2.4?

**Comment by crohart** [comment]
Yeah!!! it works now - I'm on HAProxy version 2.4.15-7782e23

**Comment by capflam** [comment]
I guess we can close this issue and consider it as fixed. I don't know if the issue is solved for the 2.2.X. at worst, it can be reopened if necessary. Thanks !
